fish-rust/src/builtins/string/join.rs

```diff
@@ -0,0 +1,106 @@
+use super::{string_error, Arguments, StringError, StringSubCommand};
+use crate::builtins::shared::{
+    io_streams_t, BUILTIN_ERR_ARG_COUNT0, STATUS_CMD_ERROR, STATUS_CMD_OK, STATUS_INVALID_ARGS,
+};
+use crate::ffi::parser_t;
+use crate::wchar::{wstr, L};
+use crate::wgetopt::{wopt, woption, woption_argument_t::*};
+use crate::wutil::wgettext_fmt;
+
+pub struct Join<'args> {
+    quiet: bool,
+    no_empty: bool,
+    pub is_join0: bool,
+    sep: &'args wstr,
+}
+
+impl Default for Join<'_> {
+    fn default() -> Self {
+        Self {
+            quiet: false,
+            no_empty: false,
+            is_join0: false,
+            sep: L!("\0"),
+        }
+    }
+}
+
+impl<'args> StringSubCommand<'args> for Join<'args> {
+    const LONG_OPTIONS: &'static [woption<'static>] = &[
+        wopt(L!("quiet"), no_argument, 'q'),
+        wopt(L!("no-empty"), no_argument, 'n'),
+    ];
+    const SHORT_OPTIONS: &'static wstr = L!(":qn");
+
+    fn parse_opt(&mut self, _optarg: Option<&wstr>, c: char) -> Result<(), StringError> {
+        match c {
+            'q' => self.quiet = true,
+            'n' => self.no_empty = true,
+            _ => return Err(StringError::UnknownOption),
+        }
+        return Ok(());
+    }
+
+    fn take_args(
+        &mut self,
+        optind: &mut usize,
+        args: &[&'args wstr],
+        streams: &mut io_streams_t,
+    ) -> Option<libc::c_int> {
+        if self.is_join0 {
+            return STATUS_CMD_OK;
+        }
+
+        let Some(arg) = args.get(*optind).copied() else {
+           string_error!(streams, BUILTIN_ERR_ARG_COUNT0, args[0]);
+           return STATUS_INVALID_ARGS;
+       };
+        *optind += 1;
+        self.sep = arg;
+
+        STATUS_CMD_OK
+    }
+
+    fn handle(
+        &mut self,
+        _parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optind: &mut usize,
+        args: &[&wstr],
+    ) -> Option<libc::c_int> {
+        let sep = &self.sep;
+        let mut nargs = 0usize;
+        let mut print_newline = true;
+        for (arg, want_newline) in Arguments::new(args, optind, streams, true) {
+            if !self.quiet {
+                if self.no_empty && arg.is_empty() {
+                    continue;
+                }
+
+                if nargs > 0 {
+                    streams.out.append(sep);
+                }
+
+                streams.out.append(arg);
+            } else if nargs > 1 {
+                return STATUS_CMD_OK;
+            }
+            nargs += 1;
+            print_newline = want_newline;
+        }
+
+        if nargs > 0 && !self.quiet {
+            if self.is_join0 {
+                streams.out.append1('\0');
+            } else if print_newline {
```

Comment:
> `print_newline` is false only in a really special case (when a final newline is missing
from stdin). Let's give it a better name (everywhere) to communicate that it only applies
to the last line. maybe `print_trailing_newline` or `terminate_last_line`.

Sure.

```suggestion
        let mut print_trailing_newline = true;
        for (arg, want_newline) in Arguments::new(args, optind, streams, true) {
            if !self.quiet {
                if self.no_empty && arg.is_empty() {
                    continue;
                }

                if nargs > 0 {
                    streams.out.append(sep);
                }

                streams.out.append(arg);
            } else if nargs > 1 {
                return STATUS_CMD_OK;
            }
            nargs += 1;
            print_trailing_newline = want_newline;
        }

        if nargs > 0 && !self.quiet {
            if self.is_join0 {
                streams.out.append1('\0');
            } else if print_trailing_newline {
```